From 289a75bb40fd1048ec06b3f9b741c7648960d257 Mon Sep 17 00:00:00 2001
From: Jaegeuk Kim <jaegeuk@google.com>
Date: Tue, 15 May 2018 17:22:27 -0700
Subject: [PATCH] test-file

Signed-off-by: Jaegeuk Kim <jaegeuk@google.com>
---
 run.sh | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/run.sh b/run.sh
index edc0ba1e..cc6390cd 100755
--- a/run.sh
+++ b/run.sh
@@ -7,7 +7,7 @@
 VER=f2fs
 #DEV=nvme0n1p1
 #DEV=sdb1
-DEV=md0
+DEV=sdb
 MAIN=nvme0n1p1
 MAINDEV=sdb2
 
@@ -215,8 +215,9 @@ por_fsstress()
 	_fs_opts
 
 	while true; do
-		ltp/fsstress -x "echo 3 > /proc/sys/vm/drop_caches" -X 10 -r -f fsync=8 -f sync=0 -f write=4 -f dwrite=2 -f truncate=6 -f allocsp=0 -f bulkstat=0 -f bulkstat1=0 -f freesp=0 -f zero=1 -f collapse=1 -f insert=1 -f resvsp=0 -f unresvsp=0 -S t -p 20 -n 200000 -d $TESTDIR/test &
-		sleep 10
+		ltp/fsstress -x "echo 3 > /proc/sys/vm/drop_caches" -X 1000 -r -f fsync=0 -f sync=0 -f write=10 -f dwrite=4 -f truncate=6 -f allocsp=0 -f bulkstat=0 -f bulkstat1=0 -f freesp=0 -f zero=1 -f collapse=1 -f insert=1 -f resvsp=0 -f unresvsp=0 -f link=1 -S t -p 10 -n 200000 -d $TESTDIR/test &
+		#ltp/fsstress -x "echo 3 > /proc/sys/vm/drop_caches" -X 10 -r -f fsync=8 -f sync=0 -f write=4 -f dwrite=2 -f truncate=6 -f allocsp=0 -f bulkstat=0 -f bulkstat1=0 -f freesp=0 -f zero=1 -f collapse=1 -f insert=1 -f resvsp=0 -f unresvsp=0 -S t -p 20 -n 200000 -d $TESTDIR/test &
+		sleep 20
 		src/godown $TESTDIR
 		killall fsstress
 		sleep 5
@@ -224,7 +225,7 @@ por_fsstress()
 		echo 3 > /proc/sys/vm/drop_caches
 		_fsck
 		_mount f2fs
-		rm $TESTDIR/testfile
+		rm $TETDIR/testfile
 		touch $TESTDIR/testfile
 		umount $TESTDIR
 		_fsck
@@ -544,7 +545,7 @@ reload)
 	#_error
 	_mount f2fs
 	#_fs_opts
-#	echo foo | f2fscrypt add_key -S 0x12 $TESTDIR
+	echo foo | f2fscrypt add_key -S 0x12 $TESTDIR
 	;;
 reload_ko)
 	_umount
-- 
2.14.1

